ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,393 @@
+    pub fn get_untracked_page(&self, addr: Vaddr) -> Option<(Paddr, usize)> {
+        let query_result = self.query_page(addr);
+        match query_result {
+            PageTableItem::MappedUntracked {
+                va: _,
+                pa,
+                len,
+                prop: _,
+            } => Some((pa, len)),
+            _ => None,
+        }
+    }
+}
+
+impl<M: AllocatorSelector + 'static> Drop for Kva<M> {
+    fn drop(&mut self) {
+        // 1. unmap all mapped pages.
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let range = self.start()..self.end();
+        let mut cursor = page_table.cursor_mut(&range).unwrap();
+        let flusher = TlbFlusher::new(CpuSet::new_full(), disable_preempt());
+        let tlb_prefer_flush_all = self.end() - self.start() > FLUSH_ALL_RANGE_THRESHOLD;
+
+        loop {
+            let result = unsafe { cursor.take_next(self.end() - cursor.virt_addr()) };
+            match result {
+                PageTableItem::Mapped { va, page, .. } => match TypeId::of::<M>() {
+                    id if id == TypeId::of::<Tracked>() => {
+                        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+                            // Only on single-CPU cases we can drop the page immediately before flushing.
+                            drop(page);
+                            continue;
+                        }
+                        flusher.issue_tlb_flush_with(TlbFlushOp::Address(va), page);
+                    }
+                    id if id == TypeId::of::<UnTracked>() => {
+                        panic!("Found tracked memory mapped into untracked `Kva`");
+                    }
+                    _ => panic!("Unexpected `Kva` type"),
+                },
+                PageTableItem::MappedUntracked { va, .. } => match TypeId::of::<M>() {
+                    id if id == TypeId::of::<UnTracked>() => {
+                        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+                            continue;
+                        }
+                        flusher.issue_tlb_flush(TlbFlushOp::Address(va));
+                    }
+                    id if id == TypeId::of::<Tracked>() => {
+                        panic!("found untracked memory mapped into tracked `Kva`");
+                    }
+                    _ => panic!("Unexpected `Kva` type"),
+                },
+                PageTableItem::NotMapped { .. } => {
+                    break;
+                }
+            }
+        }
+
+        if !flusher.need_remote_flush() && tlb_prefer_flush_all {
+            flusher.issue_tlb_flush(TlbFlushOp::All);
+        }
+
+        flusher.dispatch_tlb_flush();
+
+        // 2. free the virtual block
+        let allocator = M::select_allocator();
+        allocator.free(range);
+    }
+}
+
+impl Kva<UnTracked> {
+    fn reader(&self) -> VmReader<'_, Infallible> {
+        // SAFETY: Reading from untracked memory is safe.
```

Comment:
Neither reading to or writing from untracked memory is safe. The safety mark here should be:

```rust
// SAFETY: The constructor of the `Kva<Untracked>` structure has already ensured the safety of reading from the mapped physical address, and the mapping is valid.
```